### PR TITLE
Html printing cleanup

### PR DESCRIPTION
**Pull request by vadz**
I was trying to understand how pagination in HTML printing worked in order to use it in my own code and noticed a few things that seemed simple to fix, so I did it (there are quite a few more remaining though...).

There should be no changes in behaviour.

@vslavik I'm not going to request a review from you because it's not that important but any comments would still be welcome, of course. TIA!

### CODE REVIEW

**Comment by vadz** [comment]
Vaclav, sorry but I'm going to change my mind and ask you to review this, after all, as after looking at this code again and again, I realized that it was much more complicated and less clear than it could be and so decided to simplify it. The most important changes are the refactoring of `wxHtmlDCRenderer::Render()` into `FindNextPageBreak()` and `Render()` itself in e01892c669 and then getting rid of the array/vector of known page breaks in b9b6ccb804 as, IMO, it's completely unnecessary. Another thing is stopping calling `AdjustPageBreak()` in the loop and removing insistent mentions of having to do it from the comments/documentation in 63add2cd19 because, to the best of my knowledge, this can never do any good (well, anything better than calling it just once). Please let me know if I'm missing anything here, TIA!

**Review by vslavik — COMMENTED** [review]
Looks OK to me, but don't put too much weight on that: I don't really remember any of this code, it's a depressing trip down the memory lane more than anything else.

**Comment by vslavik** [comment]
> and then getting rid of the array/vector of known page breaks in b9b6ccb as, IMO, it's completely unnecessary. 

I'm not 100% sure of that, but I trust you tested the change. I *think* the motivation for that was the degenerated case of multiple subsequent such breaks and trying to prevent insertion of empty pages between them. I don't know what the *correct* behavior is supposed to be, though: the spec says the `<div>` that uses `page-break-before` cannot be empty, but how should something like
```html
<div style="page-break-before:always">
  <div style="page-break-before:always">
    <div style="page-break-before:always">
      on snap
    </div>
  </div>
</div>
```
behave is unclear to me.

> Another thing is stopping calling AdjustPageBreak() in the loop and removing insistent mentions of having to do it from the comments/documentation in 63add2c because, to the best of my knowledge, this can never do any good

Again just a hypothesis, but I think the motivation was that *if* `AdjustPagebreak()` could move the pagebreak in either direction, then multiple passes would be necessary. But as long as it's moved only backwards (as is the case with the existing implementations), the 2018 me doesn't see why single pass wouldn't suffice either.

File: src/html/htmprint.cpp [hunk 1]
```diff
@@ -199,7 +199,6 @@ wxVector<wxHtmlFilter*> wxHtmlPrintout::m_Filters;
 
 wxHtmlPrintout::wxHtmlPrintout(const wxString& title) : wxPrintout(title)
 {
-    m_NumPages = INT_MAX;
```
[the remark this answers is not in the document]
Comment (vadz) [reply src/html/htmprint.cpp]: AFAICS it wasn't, which is one of the reasons why this code was rather confusing. Note that I still preserved returning `INT_MAX` in `GetPageInfo()`, which seems to be the only thing it was used for, even though I'm not really sure if `GetPageInfo()` can be called before `OnPreparePrinting()` (wxGTK code is the only place where it's not totally clear where this happens).

**Comment by vadz** [comment]
I've tested (with the new unit test and in the `html/printing` sample) with the things I could think of, but I didn't think about your example above, for example. Testing it now, I think it does behave correctly: as written, there are no page breaks at all because there is already an implicit page break at position 0. If I add something before it, there is similarly just a single page break because nested divs are at position 0 of their parent and so don't create page breaks, i.e. this test passes:
```
    pr.SetHtmlText
       (
            "Something"
            "<div style=\"page-break-before:always\">"
                "<div style=\"page-break-before:always\">"
                    "<div style=\"page-break-before:always\">"
                        "whatever"
                    "</div>"
                "</div>"
            "</div>"
       );
    CHECK( CountPages(pr) == 2 );
```

I think this is the expected behaviour, isn't it?

As for moving page breaks forward: I hadn't found any version (and I looked at a few different iterations of this code all the way back to the time it was written) where it worked like this. I.e. it only ever moved backwards/higher and this also makes sense to me.

Thanks for reviewing and I'll merge this soon if there are no other comments.

BTW, the next thing I plan to do is to implement [`page-break-inside: avoid`](https://developer.mozilla.org/en-US/docs/Web/CSS/page-break-inside) support. AFAICS it should be simple using `SetCanLiveOnPagebreak()`, but I haven't tested it yet.

**Comment by vslavik** [comment]
> I think this is the expected behaviour, isn't it?

I'd say so, yes.

> Thanks for reviewing and I'll merge this soon if there are no other comments.

There aren't, and the above ones were meant just as explanatory for things that (rightly) made no sense to you.

Thanks for cleaning up this part!

**Pull request merged**